### PR TITLE
B 33971/global setup teardown

### PR DESCRIPTION
**Pull request by sam-falvo**
We have a body of integration test suites currently running sequentially.  We need these to run in parallel.  However, each setUp procedure attempts to establish some global resource; e.g., a Cassandra schema.  A batch of five schema restores running all at once leaves all five tests in a failure state, since Cassandra insists that the versions are all out of date.

This PR adds a feature to Whiskey which allows us to pull that common initialization code out into a single globalSetUp that is run for the _entire_ batch of test suites.  This way, the Cassandra code executes exactly once., regardless of parallelism used.

### CODE REVIEW

File: lib/run.js [hunk 1]
```diff
@@ -62,6 +63,24 @@ function TestRunner(options) {
...
+  this._global_setup_procedure = ignored;
+  this._global_teardown_procedure = ignored;
+
+  var global_setup_teardown_file = options['global-setup-teardown'] || null;
+  if (global_setup_teardown_file) {
+    if (global_setup_teardown_file.charAt(0) !== '/') {
+      global_setup_teardown_file = path.join(process.cwd(), global_setup_teardown_file);
+    }
+
+    var gst_exports = require(global_setup_teardown_file);
```
Comment (kans) [inline lib/run.js]: this var will get hosted to the top of the function since JS is function scoped.  You may want to move the declaration up.

Comment (sam-falvo) [inline lib/run.js]: Just to clarify, do you mean doing this?

var gst_exports;

...etc...
if (global_setup_teardown_file) {
  ...
  gst_exports = require(global_setup_teardown_file);
  ...
}

?

File: lib/run.js [hunk 2]
```diff
@@ -62,6 +63,24 @@ function TestRunner(options) {
...
+    if (gst_exports.hasOwnProperty(constants.GLOBAL_SETUP_FUNCTION_NAME)) {
+      this._global_setup_procedure = gst_exports.globalSetUp;
```
Comment (kans) [inline lib/run.js]: Do we care if this is set to something stupid (ie, not a function)?

gst_exports.hasOwnProperty potentially differs from Object.prototype.hasOwnProperty- also, we maybe want to go up the prototype chain here?  Also, if you are keeping underscore, check out _.has

File: lib/run.js [hunk 3]
```diff
@@ -43,6 +43,7 @@ exports.exitCode = 0;
 exports.processRunner = null;
 
 function TestRunner(options) {
+  var ignored = function(ignored_test, ignored_assert) {};
```
Comment (kans) [inline lib/run.js:46]: Does this function need to call ignored_test.complete or done or something for whiskey's sake?

File: lib/run.js (hunk 2 above)
Comment (sam-falvo) [inline lib/run.js]: I do expect it to be a function.  I'll look into adding a check for that.

No, we do not want to go up the prototype chain.  gst_exports must either reference the global setup procedure in the global setup/teardown file, or it doesn't.  There can be no interpolation of intent here, or else things get awfully hard to debug.

File: lib/run.js (hunk 3 above)
Comment (sam-falvo) [inline lib/run.js:46]: Nope; tests show that it works fine without it.

File: lib/run.js (hunk 1 above)
Comment (kans) [inline lib/run.js]: yep.

**Comment by kans** [comment]
@sam-falvo recent changes look fine.  Just needs a squash

**Comment by sam-falvo** [comment]
Obsoleted by the global_setup_teardown branch PR.  The +1 from this PR will apply to the new branch however.

**Pull request closed without merging**